### ISSUE TITLE
Rotating device causing index out of range.

### ISSUE THREAD

**Issue by GeorgeA93** [ephread/Instructions#63]
The following is my implementation of the coachMarksForIndex data source function.
```
func coachMarksController(coachMarksController: CoachMarksController, coachMarksForIndex index: Int) -> CoachMark {
            let instruction = self.instructions[index]
            return coachMarksController.coachMarkForView(instruction.pointOfInterest)
}
```
As you can see I am accessing an array to determine the point of interest for each CoachMark.

This works as expected the first time you start the coachMarksController. However, during a rotation of the device (iPad Pro) the data source method fires, passing '-1' as the index. Obviously causing the index out of range exception. 

Please note I am rotating the device after the coachMarksController finishes the animation; there are no coachMarks on screen.

I originally thought I was doing something wrong, however I am slightly confused as to why the data source function is firing when there are no coachMarks on screen.

I then realised that after running the coachMarkController once, I was not able to run it again. (I soon saw that you have fixed this #58 but not released it?)

Any ideas?

**Comment by ephread** [comment]
Hello @GeorgeA93 thanks for reporting the issue! Indeed, #58 hasn't been released yet. Obviously, as you suspect, the dataSource methods aren't supposed to be called if coach marks are not showing. I'll look into it.

**Comment by GeorgeA93** [comment]
Thanks for your reply. For anyone interested I am using the following work around to prevent the error:

```
func coachMarksController(coachMarksController: CoachMarksController, coachMarksForIndex index: Int) -> CoachMark {
    if index < 0 || index >= self.instructions.count {
            return coachMarksController.coachMarkForView()
    }
    let instruction = self.instructions[index]
    return coachMarksController.coachMarkForView(instruction.pointOfInterest)
}
```

This issue won't affect most users of this library because they are probably using a switch statement instead of accessing an array by index. (Assuming they have correctly implemented a default case in their switch statement)

**Comment by ephread** [comment]
Great! 👏 

By the way, I've suggested a workaround in the [gitter chat](https://gitter.im/ephread/Instructions?at=57acde3cf7f11a7936b6d6e3) for #58. Both issues will be fixed by 0.5.0 which should comme out before the end of august.

**Comment by ephread** [comment]
Fixed by the big refactoring (c2cd243..7425eb7).
